synapse/storage/state.py

```diff
@@ -356,6 +356,85 @@ def get_member_split(self) -> Tuple["StateFilter", "StateFilter"]:
 
         return member_filter, non_member_filter
 
+    def approx_difference(self, subtrahend: "StateFilter") -> "StateFilter":
+        """
+        Returns a state filter which represents `self - subtrahend`.
+
+        The resultant state filter MUST admit all state events that are admitted
+        by only this filter (`self`) and not `subtrahend`. (1)
+        The resultant filter MAY be an over-approximation: the resultant state
+        filter MAY additionally admit other state events.
+
+
+        Formally, if the set of state events admitted by a state filter F are
+        written as E(F), then the resultant state filter bears this property:
+
+            E(difference(self, subtrahend)) ⊇ E(self) ∖ E(subtrahend)
```

Comment:
I can make the left hand side bigger while still preserving this property in two ways:

1. Add elements in neither E(self) nor E(subtrahend)
2. Add elements in E(subtrahend)

I think the prose below suggests that the approximation might be inexact due to 2 but not due to 1. Is that correct?

If so, I suggest replacing this with 

```suggestion
            E(self) \ E(subtrahend) ⊆ E(approx_difference) ⊆ E(self)
```

Reply:
And again, if so, suggest "MAY additionally admit other state events" be altered, e.g. to "MAY admit some events from `subtrahend`"

Reply:
You're right.

I'm not actually 'interested' in the stronger property that you write — the code using this function does not depend on it and is happy with `StateFilter.all()` as a return value (for example) — though equally just returning `self` is also a satisfactory 'lazy, technically correct' answer.

For the sake of potential re-use in the future, I think what you are writing is sensible, so I'll try and incorporate it.